Fix for update products via csv file (fix for 22028)

mtwegrzycki opened the pull request:
### Description (*)
It is fix for issue #22028

In query for selecting ids to indexing after import products, I changed "between" into "in",
Before change, when products were updated via csv file import in Admin panel, to indexing process were selected all products from range between the smallest and the biggest products id from csv file. When the ids were from wide range, it caused selecting a huge amount of data and attempt to inserting them to catalog_product_index_price_temp. If data amount was to big, it caused error
After change, for indexing process are selected only products from csv file.

Example:
Before change, when we try to update two products with id 30 and 31, only two products were selected to indexing. 
When products have ids 30 and 50, it caused, that 21 products were selected to indexing (All products with ids between 30 and 50.

After change, in both cases described above, only two products will be selected to indexing.

### Fixed Issues (if relevant)
No related issues found.

### Manual testing scenarios (*)
The same scenario can be applied to reproduce bug and check fix:
1. Prepare csv file to update products (2 products or more). (This csv should contain products with the lowest and the highest id - 1 and 6000)
2. In Admin panel go to: System->Import and choose Product and Add/Update
3. Upload file
4. Check Data
5. Import csv file (error should be displayed now)

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

mtwegrzycki committed:
Fix for update products via csv file (fix for 22028)

Fix for update products via csv file (fix for 22028)

Fix for update products via csv file (fix for 22028)

magento-cicd2 commented:
[![CLA assistant check](https://cla.dev.magento.com/pull/badge/not_signed)](https://cla.dev.magento.com/magento/magento2?pullRequest=22575) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=22575) before we can accept your contribution.<br/><hr/>**Mateusz Wegrzycki** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

mtwegrzycki commented:
@orlangur , it is the same pull request like in #22032 , but I corrupted my repository and PR was closed automatically.

orlangur reviewed (approved):
LGTM!

@mtwegrzycki please make sure commit is linked to your GitHub user and CLA is signed.

mtwegrzycki commented:
Okay, I fixed issue with CLA.

magento-engcom-team commented:
@mtwegrzycki thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

mtwegrzycki commented:
I can see, that there any integration errors in Travis. But it does not look like related to my code changes.
Please, let me know, if there are required any additional actions from me.

orlangur reviewed (approved):
@mtwegrzycki true, build failure seems to be random.

magento-engcom-team commented:
Hi @orlangur, thank you for the review.
[ENGCOM-5028](https://jira.corp.magento.com/browse/ENGCOM-5028) has been created to process this Pull Request

Nazar65 commented:
Hi @mtwegrzycki thank you for Pull request, can you please explain? why you add your changes to else condition, ? becouse if you can see there is a validation `count($entityIds) > 1` so if you have 1 entityIds then it switch in else and this looks like `where e.entity_id in (1);`
```
if (count($entityIds) > 1) {
                $select->where(sprintf('e.entity_id BETWEEN %s AND %s', min($entityIds), max($entityIds)));
            } else {
                $select->where('e.entity_id = ?', $entityIds);
                $select->where('e.entity_id IN(?)', $entityIds);
            }
```

mtwegrzycki commented:
Hi @Nazar65,
I didn't add any changes in else condition. I added IN after BETWEEN in if condition and removed else condition. 
I checked and it doesn't matter if we use IN or equals for one id(item) in MySQL.
Part of code in your message is not the same as my PR. My change looks like this:
```
if (count($entityIds) > 1) {
                $select->where(sprintf('e.entity_id BETWEEN %s AND %s', min($entityIds), max($entityIds)));
                $select->where('e.entity_id IN(?)', $entityIds);
            }
```

Please, let me know, if you need more detailed information.

mtwegrzycki commented:
Okay, now I see the bug in my code.
If id will be equaled 1, SQL query will be not executed.
I should remove if condition and leave only two where queries.

Nazar65 commented:
@mtwegrzycki yeap. if we have one ID they not be indexed.

mtwegrzycki commented:
Okay, I fixed it. :D

Nazar65 commented:
@orlangur can you please review ? and should we use BETWEN ? as IN works perfectly with one entity id.

orlangur commented:
@Nazar65 regarding between check in comment I referenced.

magento-engcom-team commented:
Hi @orlangur, thank you for the review.
[ENGCOM-5028](https://jira.corp.magento.com/browse/ENGCOM-5028) has been created to process this Pull Request

stoleksiy commented:
:heavy_check_mark: QA Passed

hatimeria-artur-jewula commented:
If you are adding "IN" why the need to keep "BETWEEN"? I assumed the between statement was introduced in case a large number of items need reindexing. As the query itself is pretty long (from what I can see about 11k characters) it could cause fail due to batch size limit. But instead of between statement the better solution in my opinion is to chunk entity ids array and execute 2 or 3 queries.

orlangur commented:
@hatimeria-artur-jewula this suggestion came from https://github.com/magento/magento2/pull/19878#discussion_r243217689
